### DISCUSSION TITLE
🗺 Roadmap

### DISCUSSION THREAD

**Discussion by pirate** [ArchiveBox/ArchiveBox#120] [General]
The plans for the next versions ahead can be seen here:

_**[`The Roadmap`](https://github.com/ArchiveBox/ArchiveBox/wiki/Roadmap)**_

Discuss your questions, comments, and concerns about large upcoming changes and high-level design direction in this thread.

*Please DO NOT use this thread to discuss general bugs, small feature requests, or other things unrelated to long-term high-level roadmap direction or major architectural decisions.*  
*(I will hide off-topic comments and direct you to open separate issues instead)*

---

To see how the roadmap has been executed on so far, read these pull requests:

 - ✅ [v0.2.x](https://github.com/ArchiveBox/ArchiveBox/tree/483a3bef9e2b1a7b80611947a3be99b0cf4f9959) 
 - ✅ [v0.3.x](https://github.com/ArchiveBox/ArchiveBox/pull/197)
 - ✅ [v0.4.x](https://github.com/ArchiveBox/ArchiveBox/pull/207)
 - ✅ [v0.5.x](https://github.com/ArchiveBox/ArchiveBox/pull/552)
 - ✅ [v0.6.x](https://github.com/ArchiveBox/ArchiveBox/pull/680)
 - ✅ [v0.7.x](https://github.com/ArchiveBox/ArchiveBox/pull/721)
 - 🛠 [v0.8.x](https://github.com/pirate/archivebox/releases)

*[See all releases...](https://github.com/pirate/archivebox/releases)*

---

**IMPORTANT:** *Please don't work on implementing any of the things in the roadmap without opening an issue to discuss it first, work is already in progress for big parts of these designs, and I may have to reject your PR if it doesn't align with the existing work!*

**Comment by pigmonkey** [comment]
I think it's important to maintain the simplicity of being able to use ArchiveBox as a simple command line tool, which makes me shy away from something like redis. I don't want to have to spin that up on my laptop just to archive my bookmarks. I also wonder if it might be worth considering making dramatiq an optional dependency.

The `depth` flag for crawling will be neat. Further down the line it might be useful if there was a way to store whitelisted domains for each entry to pass to the crawler, similar to wget's `--domain`.

**Comment by pirate** [comment]
Redis, dramatiq, etc could all be optional dependencies used only for the web GUI.

**Comment by mbreedlove** [comment]
@pirate do you have any high-level goals for `archivebox server`? I am interested in helping to define requirements for the front-end web app.

Also, can you document the differences between the `add` and `crawl` commands? It seems like `crawl` operates on a list, while `add` operated on a single URL. If so, I think crawl would be better named `add-list`, or something else that indicates that it behaves like `add` but on a list.

**Comment by pirate** [comment]
@mbreedlove I haven't gotten around to fully spec'ing out a frontend yet, but as inspiration you can look at https://github.com/kanishka-linux/reminiscence and webrecorder.io which are very similar to what I want ArchiveBox server to end up as (with the addition of some federated archive features so people can share resources with eachother).

As for crawl vs add, I'm still on the fence about having crawl at all, I think I may end up just going with `add --depth=[n]` to specify whether you want to crawl or add just one page, where depth defaults to 0 and just adds the one page.  Simplicity is king, so if I can make it intuitive with just `add` that would be ideal.  Maybe `crawl` can just be an alias for `add --depth=1`.

**Comment by JoshuaMorris** [comment]
It would be awesome to have the option to install from git/select from contributed themes for the frontend. I'm interested in contributing

**Comment by pirate** [comment]
Sure! If you want to contribute some themes right now you're welcome to pr some new files into an `archivebox/templates/<theme-name>` directory.  Then users can select them using the [`TEMPLATES_DIR`](https://github.com/pirate/ArchiveBox/wiki/Configuration#TEMPLATES_DIR) config option.

**Comment by anarcat** [comment]
it would be really nice if archivebox was a more "normal" python program that is shipped on pypi.org and can be installed with pip or at least `setup.py install`. it would make debian packaging easier (if not just "possible" at all). as things stand now, i'd really like to start using this but i'm hesitant because of the lack of standard installation procedures and maintenance burden...

should i file that as a separate feature request? :)

**Comment by pirate** [comment]
@anarcat this is [high on our priority list](https://github.com/pirate/ArchiveBox/issues/177) already, if you notice I mention the pip package in a few places in the roadmap comment at the top.  The planned api is to allow users to simply `pip install archivebox` to get the entire program working right out of the box.

The only reason I haven't already gotten around to it yet is because I want to release the new django db schema + migration system so that I can push updates to people via pip without breaking their existing archives by accident.  (the migrations system will allow us to do forwards and backwards data migrations safely by using a SQLite db instead of relatively brittle json files).

**Comment by anarcat** [comment]
On 2019-03-08 10:25:31, Nick Sweeting wrote:
> @anarcat this is high on our priority list already, if you notice I mention the pip package in a few places in the roadmap comment at the top.  The planned api is to allow users to simply `pip install archivebox` to get the entire program working right out of the box.

Speaking from a Debian packaging perspective, that would make my life
much easier. Do let me know when that's done and I'll see what I can do
about shipping this in Debian. It's too late for the next stable release
but I'm sure it will be trivial to backport...

Packaging progress in Debian is being tracked here:

https://bugs.debian.org/924040

**Comment by pirate** [comment]
WOAH, life-goal of mine to have a package in Debian! Thanks for starting that process, I'm of course happy to help maintain the debian port once the pip package is out.  I'll be sure to keep you posted on the pip packaging process, although it won't be super soon since that new backend has a lot of complexity and ArchiveBox is only my spare-time project (expect somewhere in the 4-6 month range).

**Comment by 007** [comment]
Feature request for roadmap: Docker nativity

Please make sure that whatever requirements this project has can be packaged into a container, and that one or two mounts can be provided as a durable backing store.

I regularly break things on my server(s) and need to reinstall either a full OS or some subset. Sometimes it's a bad driver, sometimes I end up with a Python package installed globally because I forgot a `venv`. I've started moving my own works to docker image + data directory to prevent lots of trouble like this. It also helps with multiple versions when I'm developing, since I know with certainty that exactly what is committed is what I'm building - no extra / missing `requirements.txt` or `apt-get install` packages.

**Comment by pirate** [comment]
We don't intend to break docker support by any means, what you described is exactly how I do things on my servers too.  Also in case you're not aware, we already have an existing Docker + ArchiveBox setup documented, you can check it out here: https://github.com/pirate/ArchiveBox/wiki/Docker

**Comment by pirate** [comment]
For everyone who's interested, we're starting the pip packaging process now, and you can follow progress here: https://github.com/pirate/ArchiveBox/issues/177

The good news is that dependencies will be entirely handled with python/pip going forwards. We'll no longer ask users to `apt`/`brew` install anything before runing ArchiveBox, just   
`pip install ArchiveBox` or   
`docker run -d -v "./data:/data -p 8089:80 archivebox`   
and it's ready to go.

**Comment by anarcat** [comment]
that's awesome @pirate! 

the Docker setup looks pretty good already, but something that might make it better is if automated builds would be enabled in an officially-sanctioned repo on hub.docker.com. Right now, if we [search for archivebox](https://hub.docker.com/search?q=archivebox&type=image) there we find only two unrelated images.

people have a tendency to use those sometimes, maybe even blindly, so it's a good idea to claim a space there when you can.

**Comment by pirate** [comment]
@anarcat done: https://hub.docker.com/r/nikisweeting/archivebox

I've also updated the [Docker](https://github.com/pirate/ArchiveBox/wiki/Docker) wiki page to reference the official image and use the new tag to download it.

I'll also go through the process to get it listed on: https://github.com/docker-library/official-images and https://github.com/docker-library/docs with a proper icon and verification of project ownership.

**Comment by anarcat** [comment]
awesome, you're on fire! :)

**Comment by sbrl** [comment]
That's cool, @pirate! How will that work with Chromium though? Won't you have to apt install that?

**Comment by pirate** [comment]
@sbrl `pyppeteer` handles the Chromium bundling automatically, so we won't have to `apt` install it.

<img width="880" alt="image" src="https://user-images.githubusercontent.com/511499/54464738-40b2ed80-474e-11e9-9c71-f5d6955b9a2e.png">

**Comment by sbrl** [comment]
@pirate Wow. Can't wait! :D

**Comment by Marthisdil** [comment]
You may wish to add an option when --depth is used

If I wanted to make an archive of a whole website (for example), including all sub links, etc, and I set depth to either infinite (or some arbitrarily high number like 9999), an option to keep archivebox on the main site's URL only would be awesome (like --stayonsite).

That way, if it comes across a link that takes you off of the site, it doesn't go down that rabbit hole.

Example: echo 'https://www.site.com' | ./archive --depth 9999 --stayonsite      - will archive down to a depth of 9999 on site.com *only* and won't leave the hierarchy.

**Comment by pirate** [comment]
I think we can add that behind a `--mirror` flag, like wget has @Marthisdil.

**Comment by Marthisdil** [comment]
@pirate sounds good!  Thanks for the hard work!  Looking forward to it!

**Comment by Kovah** [comment]
Currently Archivebox seems to update existing entries when adding a new URL, couldn't find an option of flag to turn this off.
Does `archivebox add  https://example.com/some/page` mean that this command will only add this new URL without doing enything else? Would be awesome!

Edit: nevermind, just found the `ONLY_NEW` option.

**Comment by pirate** [comment]
FYI for everyone following this issue, I've put some major work in over the last week, and the new Django release is almost done:

 - Final stable pre-django release: https://github.com/pirate/ArchiveBox/pull/197
 - First unstable beta django release: https://github.com/pirate/ArchiveBox/pull/207

The final pre-django release (v0.3.0) is mostly a stability/integrity release that makes sure corrupted indexes are never written to disk, and that the indexes are in the correct format for Django to import.  It has a bunch of bugfixes as well, but no major API/internal schema changes other than the switch from untyped dicts storage to mypy-enforced typed dataclasses.

The Django release (v0.4.0) is the big one that ya'll have been waiting for, it includes an entirely new CLI UX where you `cd` into the data directory and run `archivebox init` to create/import an archive (sort of like `npm init`), instead of using `env OUTPUT_DIR` or defaulting to `./output`.  This release is the one that finally cleans up a lot of the rough edges in the ArchiveBox experience, and moves us over from a brittle, untyped json data storage system to SQLite + optional json/csv exports.

Here's a sneak-peek of some of the new v0.4.0 CLI (this is [already working](https://github.com/pirate/ArchiveBox/pull/207), check out the `django` branch to test it, see [documentation here](https://github.com/pirate/ArchiveBox/wiki/Roadmap)!):
```bash
$ archivebox init
$ archivebox add https://example.com/some/page/to/archive
$ archivebox add --depth=1 https://example.com/some/rss/feed/to/archive.xml
$ archivebox add --depth=1 < ~/downloads/bookmarks_export.html
$ archivebox list --sort=timestamp --filter='.*github.com.*' --csv=timestamp,base_url,is_archived
$ archivebox update --resume=1554263415
```

I encourage you to check out both PRs linked above and leave your feedback as comments.  I'll be pushing the final round of commits and testing both releases over the next 1-2 weeks.

**Comment by sbrl** [comment]
So the upgrade path will be v0.3.0 -> v0.4.0?

Also, love the new globe map in the title :D

**Comment by pirate** [comment]
Yup, you can also just go straight to v0.4.0 from v0.2.4. The middle release is just so I have a nice stable version that’s before the huge Django refactor in case people don’t like Django or we have to go back for some reason.  It’s not a required step for users to install it before the Django release.

**Comment by sbrl** [comment]
Ah, I see! Thanks for the clarification.

I'll test them both out as they are released :-)

**Comment by lachlansimpson** [comment]
Checked out the django branch, but I've obv done something wrong - I'm not getting much love:

```
ubuntu@archive-box:~ $ pip3 install archivebox
Collecting archivebox
  Could not find a version that satisfies the requirement archivebox (from versions: )
No matching distribution found for archivebox
ubuntu@archive-box:~ $cd /opt/ArchiveBox_django 
ubuntu@archive-box:/opt/ArchiveBox_django$ bin/archivebox init
Traceback (most recent call last):
  File "bin/archivebox", line 5, in <module>
    from .cli.archivebox import main
ModuleNotFoundError: No module named 'archivebox'
```

Any hints? This is a fresh ubuntu 18.04 installation with nothing else installed.

**Comment by pirate** [comment]
`pip3 install -e .` (it’s not on pypi until it’s released so you have to install from the local repo dir instead of pypi.

**Comment by lachlansimpson** [comment]
dataclasses are python 3.7? For ubuntu users with python 3.6, `pip3 install dataclasses`

**Comment by lachlansimpson** [comment]
```
    ! Failed to archive link: ValueError: Tried to parse invalid date! 2019-04-07T05:44:39.227520                                       
....
  File "/opt/ArchiveBox_django/archivebox/legacy/util.py", line 152, in typechecked_function
    return func(*args, **kwargs)
  File "/opt/ArchiveBox_django/archivebox/legacy/util.py", line 404, in parse_date
    raise ValueError('Tried to parse invalid date! {}'.format(date))
ValueError: Tried to parse invalid date! 2019-04-07T05:44:39.227520
```

**Comment by pirate** [comment]
Thanks @datakid I added dataclasses as a dependency in `setup.py`, I've also fixed that dateparse issue on v0.4.0.

**Comment by Marthisdil** [comment]
I like the --depth modifier.  However, how would you use this to mirror a whole site (without the spider moving OFF the site if the site had a link to another site)?

**Comment by pirate** [comment]
@Marthisdil the `--mirror` option will handle archiving a site without leaving the domain (same as wget's behavior): https://github.com/pirate/ArchiveBox/wiki/Roadmap#--mirror

For everyone else, 😢 **please read [this issue](https://github.com/pirate/ArchiveBox/pull/207#issuecomment-494107553), there are some major problems with the ArchiveBox security model around how we deal with executing JS in archived content** without allowing it to XSS the entire archive.  This is a pretty major blocker as I'm not comfortable pushing a new version until it's "secure by default" and doesn't risk exposing people's feeds.  I will try to put some time in, but this month and next are going to be very busy for me so PRs helping with the JS security would be very appreciated! 

For more details on the security and JS issues, see this PR comment: https://github.com/pirate/ArchiveBox/pull/207#issuecomment-494107553
I'll also be updating the readme soon to responsibly disclose the current attack vectors and outline our approach to fixing it (stripping and preventing any JS execution via static archive html and titles and only allowing JS execution via WARC replay).

This architectural problem happened because I originally designed ArchiveBox to only archive public feeds and publicly accessible pages, where leaking the list of pages in the archive was not a concern.  When more people started using it for private content, it made sense to support that use case out-of-the-box.  But now I need to go back and re-assess the assumptions that were made by the original design, and do it properly. I'll have to sacrifice some of the nice UX of having easily accessible archived files on disk, but unfortunately after talking about it with a few security people over the last weak, it's clear that there cannot be any insecurity by default, even if it does make the UX easier, and we must correct flawed design assumptions ASAP.  I'll try and provide config options for people to get the old behavior back, but at this stage it looks like static archives will no longer allow executing JS anymore, breaking any interactive pages and forcing people to use a dedicated WARC replayer if they need that ability.  PDFs, Screenshots, and all the other methods will remain unaffected, these changes will only disable JS for the primary wget static html archive.

**Comment by pirate** [comment]
**2019/07 Update:** I haven't forgotten about this, ArchiveBox is still my heart and soul and will remain my primary side-project for the long-term future. I'm just still swamped by my day job and a little daunted by the scope of the security fixes needed to get the next version ready. All JS-stripping / security PRs are extremely welcome in the meantime!  

I'll keep posting updates regularly on status, if you don't want the email notifications you're welcome to mute this issue and check back in a couple months.

<img src="https://user-images.githubusercontent.com/511499/60913191-e128a280-a254-11e9-89f2-0a1fcb5c0916.jpg" width="500px">
(I promise this isn't me haha, my weeks have actually been crazy)

**Comment by sbrl** [comment]
I wonder if it's possible to set `Content-Security-Policy` via a HTML `<meta />` tag? If so, then in theory you could disable JS execution by just artificially inserting into the head of scraped pages :thinking:

**Comment by cbcaron15** [comment]
Feature suggestion: a browser extension similar to Archiveror whereby anytime someone visits an archived page, the extension fetches links to the archives associated with that URL and displays them within the extension along with the original timestamp, or at least a single link to the specific landing page on the UI for that URL. If a webpage responds with 404, or if the user suspects significant changes to a webpage since first visit / original archive, then such an extension would offer ultra convenience for accessing the archives. Rather than having to manually navigate to the UI and search for the URL.

**Comment by sbrl** [comment]
@cbcaron15 sounds like that would do well as a separate issue.

**Comment by kingsleytorlowei** [comment]
any update?

**Comment by pirate** [comment]
Hey all, v0.4 is officially released, check it out:

https://github.com/pirate/ArchiveBox/releases/tag/v0.4.9

We've closed over 30 tickets in this sprint, and made a bunch of major improvements to the CLI experience, data integrity, and web UI!

```bash
pip install archivebox

cd path/to/your/archive/folder

archivebox init  # doubles as the upgrade command, non-destructive & safe to run multiple times
archviebox add 'https://example.com'
archviebox add 'https://getpocket.com/users/USERNAME/feed/all' --depth=1
archivebox status
# archivebox config --set PUBLIC_INDEX=True      (False=require a login to view the index)
# archivebox config --set PUBLIC_SNAPSHOTS=True  (False=require a login to view individual snapshots)
archivebox server 0.0.0.0:8000
archivebox help
```

```bash
docker run -v $PWD:/data nikisweeting/archivebox init
docker run -v $PWD:/data nikisweeting/archivebox add 'https://example.com'
docker run -v $PWD:/data -it nikisweeting/archivebox manage createsuperuser
docker run -v $PWD:/data -p 8000:8000 nikisweeting/archivebox
```

**Comment by pirate** [comment]
I forgot to post here when we released `v0.5.3` a couple months ago.

https://github.com/pirate/ArchiveBox/releases/tag/v0.5.3

- ArchiveResult moved to SQLite3 DB for performance @cdvv7788 
- lots of assorted bugfixes and improvements courtesy of @cdvv7788 and @jdcaballerov
- new full-text search support with ripgrep and sonic courtesy of @jdcaballerov
- new `archivebox oneshot` command for downloading a single site without starting a whole collection
- new Pocket API importer courtesy of @mAAdhaTTah
- new Wallabag importer courtesy of @ehainry
- new extractor options on Add page courtesy of @BlipRanger
- new apt/deb/homebrew/pip packaging setup into separate repos under new Github Org https://github.com/ArchiveBox
- new official PPA and Docker Hub accounts https://hub.docker.com/r/archivebox/archivebox (with automatic armv7 builds courtesy of @chrismeller)
- new Snapshot grid view courtesy of @jdcaballerov 

![image](https://user-images.githubusercontent.com/511499/104231980-ad1bd800-541d-11eb-9bfb-58fdfd4ce1ef.png)

**Comment by pirate** [comment]
The latest version just released today is: `v0.5.4` with lots of bugfixes and small improvements.  The focus for this past sprint was stability and UI experience. Performance is going to be the focus of the next major release (`v0.6`).

https://github.com/pirate/ArchiveBox/releases/tag/v0.5.4

Thank you contributors who helped with this release!  
@cdvv7788, @jdcaballerov, @thedanbob, @aggroskater, @mAAdhaTTah, @mario-campos, @mikaelf

- fix migration failing due to null cmd_versions in older archives a3008c8
- Publish, minor, & major version to DockerHub and add set up CodeQL codeql-analysis.yml c5b7d9f, bbb6cc8
- fix DATABASE_NAME posixpath, and dependencies dict bug 02bdb3b, 5c7842f
- use relative imports for `.util` to fix windows import clash 72e2c7b
- fix `COOKIES_FILE` config param breaking in wget ef7711f
- Refactor `should_save_extractor` methods to accept `overwrite` parameter 5420903
- Fix issue #617 by using mark_safe in combination with format_html … 1989275
- make permission chowning on docker start less fancy, respect PUID/PGID #635
- add createsuperuser flag to server command 39ec77e
- fix files icons styling and use the db exclusively for rendering them, instead of filesystem f004058, 7d8fe66, 5c54bcc, 534ead2
- limit youtubedl download size to 750m and stop splitting out audio files 3227f54
- also search url, timestamp, tags on public index 8a4edb4
- cleanup template layout in filesystem, 
- fix trailing slash problems and wget not detecting download path 9764a8e
- add response status code to headers.json c089501
- fix singlefile path used for sonic 24e2493


<img width="80%" alt="Screen Shot 2021-01-30 at 9 53 22 p" src="https://user-images.githubusercontent.com/511499/106431311-126e5200-643b-11eb-9cd9-05c3102d9945.png">

```bash
pip install --upgrade archivebox
# or
apt update && apt upgrade archivebox
# or
brew upgrade archivebox/archivebox/archivebox
```
```bash
docker pull archivebox/archivebox:0.5.4
docker run -v $PWD:/data -it archivebox/archivebox init
docker run -v $PWD:/data -p 8000:8000 archivebox/archivebox server 0.0.0.0:8000
```

**Comment by pirate** [comment]
FYI: https://github.com/ArchiveBox/ArchiveBox/issues/1082 Attention: I'm considering removing the `archivebox oneshot` command from ArchiveBox in the next 0.8 release

**Comment by paulie-g** [comment]
Driveby comment re: ```mv``` being atomic. You can not assume this. Firstly, what we're talking about here is rename(2). You would have to verify what Python actually does under the hood. ```shutil.mv``` is unlikely to give the same guarantees as rename(2), ```os.rename``` might. Secondly, it is only atomic wrt other filesystem ops. It is absolutely not crash-safe. Furthermore, this guarantee holds within a single filesystem. And even then not *every* filesystem implementation. I'm not going to go into details, but you won't like what happens on NFS and CIFS. Bltihely using this as a mechanism to implement transactions is a bad idea. It absolutely can be done, but not sans clue. If you need transactions via this method, try to find a library that's already implemented it and signpost that it won't work on network mounts (at a minimum). If what you want is just to move files in a reasonable manner, copy to a temp first (filename.part or somesuch) and then rename (or link + unlink). Check for leftover temp files at end of process.
